sdk/core/core/src/az_json_builder.c

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <az_json_builder.h>
+
+#include <az_str.h>
+
+#include <_az_cfg.h>
+
+AZ_NODISCARD az_result
+az_json_builder_init(az_json_builder * const out, az_span_action const write) {
+  AZ_CONTRACT_ARG_NOT_NULL(out);
+
+  *out = (az_json_builder){ .write = write };
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD static az_result
+az_json_builder_write_str(az_json_builder * const self, az_span const value) {
+  AZ_CONTRACT_ARG_NOT_NULL(self);
+
+  az_span_action const write = self->write;
+  AZ_RETURN_IF_FAILED(az_span_action_do(write, AZ_STR("\"")));
```

Comment:
Maybe az_span_action should take varargs and then we could call it like this (1 call instead of 3):
az_span_action(write, AZ_STR("\""), value, AZ_STR("\""));

Reply:
In some modern languages, like C#, JavaScript, varargs is safe. But in C, it's a very error-prone. I would avoid as much as I can. In the same time, to make polymorphic functions faster, we can accept a span of spans instead of one span.